Deprecate pinning infrastructure on iOS 14 / macOS 11 (#154)
BottleRocketStudios/iOS-Hyperspace: Documentation/Guides/SSL Pinning.md

> ### SSL Pinning

Reviewer: I'm not sure if there's a hard standard on this, but I think the markdown `#` characters are supposed to be functionality equivalent to HTML header tags (e.g. `<h1>`, `<h2>`, etc.). I think I've read somewhere that it's good practice to always have one top-level "h1" level header and then go from there. So, since this is a fresh document/page, we could follow that pattern to start with a single `#` here at the top level and then fill out the rest of the titles accordingly.

```markdown
# SSL Pinning
...

## iOS 14+
...

## Older iOS Versions
...
```

Author: That's a fair point - I had basically just matched what existed already, even if it was "wrong". I went back through the documentation and applied similar formatting.

> SSL Pinning is the practice of limiting the set of server certificates your app trusts when making network requests. This occurs in addition to the trust verification the operating system performs by default. This, in combination with HTTPS, allows you to better preserve the privacy and integrity of the information you are transmitting back and worth to a backend server.

Reviewer 2: should say "back and forth", but says "back and worth"

Author: 🤦 Fixed

> SSL Pinning is the practice of limiting the set of server certificates your app trusts when making network requests. This occurs in addition to the trust verification the operating system performs by default. This, in combination with HTTPS, allows you to better preserve the privacy and integrity of the information you are transmitting back and forth to a backend server.

Reviewer 3: Minor language nitpicks.
```suggestion
SSL Pinning is the practice of limiting the set of server certificates your app trusts when making network requests. This occurs in addition to the trust verification the operating system performs by default. This, in combination with HTTPS, better preserves the privacy and integrity of the information you are transmitting back-and-forth to a backend server.
```

Hyperspace has a built-in mechanism for implementing SSL pinning which is available by default when using Swift Package Manager or Carthage, and when using the `Pinning` subspec when using Cocoapods. But, depending on your deployment target, the recommended approach to implementation of certificate pinning differs.

#### iOS 14+

If your deployment target is set to iOS 14 or above, it is recommended to use the new `NSPinnedDomains` Info.plist key. This means that you do not need to use the `TrustConfiguration` in Hyperspace, and can instead rely almost entirely on the operating system to perform the extra trust validations on server SSL certificates.

More information on how to implement this new `NSPinnedDomains` Info.plist key, as well as information around other considerations and strategies involving pinning, is available [here](https://developer.apple.com/forums/thread/672291), with the standard developer documentation on the `NSPinnedDomains` Info.plist key [here](https://developer.apple.com/documentation/bundleresources/information_property_list/nsapptransportsecurity/nspinneddomains).

#### Older iOS Versions

> On the other hand, if you have not yet dropped iOS 13 you can use the Hyperspace implementation. The basics involve creating a `TrustConfiguration` that details the certificates that your app should trust, and then utilizing a `TrustValidator` to check each incoming server SSL certificate against the previously defined configuration. To make the evaluation of certificates simpler, Hyperspace also includes a `TrustValidatingTransportService` that can be used with any `BackendService` that automates this process.

Reviewer 3: ```suggestion
If you have not yet dropped iOS 13 you can use the Hyperspace implementation. The basics involve creating a `TrustConfiguration` that details the certificates that your app should trust, and then utilizing a `TrustValidator` to check each incoming server SSL certificate against the previously defined configuration. To make the evaluation of certificates simpler, Hyperspace also includes a `TrustValidatingTransportService` that can be used with any `BackendService` that automates this process.
```

> For example, let's assume we want to pin all our connections to https://jsonplaceholder.typicode.com. At the time the app connects to that TLS network, the server will provide it's SSL certificate for validation. When pinning is enabled, the app will inspect the contents of the certificate, and if that certificate does not match one that it is looking for (even if the certificate is valid and trusted by iOS) the request will be rejected.

Reviewer 3: I would even change it to:
```suggestion
For example, let's assume we want to pin all our connections to https://jsonplaceholder.typicode.com. The server will provide its SSL certificate for validation when the app connects to that TLS network. When pinning is enabled, the app will inspect the contents of the certificate. If that certificate does not match one the app is looking for (even if the certificate is valid and trusted by iOS) the request will be rejected.
```

These bytes can then be run through a SHA-256 hash (an easy way to do this is using the `CertificateHasher` in Hyperspace) to produce your pinning hash: `XlCPXC6IrttTF9Y1887kS+efCCf3uFjHW6D1TUI9f+Q=`. This hash can then be passed into a `DomainConfiguration` using `.init(domain:pinningHashes:)` or `.init(domain:encodedPinningHashes:)` (the latter of which is Base64 encoded). Please note that the hash required for the pinning implementation in Hyperspace is not identical to the one that the `NSPinnedDomains` key uses in iOS 14+.

Once the `TrustConfiguration` has been created with a `[DomainConfiguration]`, the final step is to tell Hyperspace how to validate this configuration at request-time. This can be done by manually utilizing a `TrustValidator` at the appropriate time, and then by calling `validator.handle(challenge:,handler:)`. In the vast majority of cases though, it is simpler to use the provided `TrustValidatingTransportService` which builds upon `TransportService` to automatically provide this functionality when the app receives an authentication challenge through the `URLSessionDelegate` for an appropriate domain.

> 

Reviewer: After having explained all of the components, it might be good to provide a basic code sample that ties everything together (I'm not sure if this is 100% accurate or could be simplified):

```swift
let pinningHash = "XlCPXC6IrttTF9Y1887kS+efCCf3uFjHW6D1TUI9f+Q="
let domainConfiguration = TrustConfiguration.DomainConfiguration(domain: "typicode.com", enforced: true, encodedPinningHashes: [pinningHash])
let trustConfiguration = TrustConfiguration(domainConfigurations: [domainConfiguration])
let validatingService = TrustValidatingTransportService(trustConfiguration: trustConfiguration)
let backendService = BackendService(transportService: validatingService)

let request = Request(...)
backendService.execute(request) {
    ...
}
```

> Once the `TrustConfiguration` has been created with an `[DomainConfiguration]`, the final step is to tell Hyperspace how to validate this configuration at request-time. This can be done by manually utilizing a `TrustValidator` at the appropriate time, and then by calling `validator.handle(challenge:,handler:)`. In the vast majority of cases though, it is simpler to use the provided `TrustValidatingTransportService` which builds upon `TransportService` to automatically provide this functionality when the app receives an authentication challenge through the `URLSessionDelegate` for an appropriate domain.

Reviewer 2: Should be "created with a `[DomainConfiguration]`," instead of "created with an `[DomainConfiguration]`"

Author: I was going back on forth on that, between 'an array of `DomainConfiguration`' and 'a `DomainConfiguration` array'.  Will update.

Once this is in place you should be able to observe the effects of pinning, for example using a proxy tool like Charles. When the proxy is not running, the certificate should be the "real" one and any requests you make to the pinned domain should succeed. Turning on the proxy will instead use a different proxy certificate, which will cause any requests to the pinned domain to return an error, similar to the one below:

``` 
failure(TransportFailure(error: Hyperspace.TransportError(code: Hyperspace.TransportError.Code.cancelled, underlyingError: Optional(Error Domain=NSURLErrorDomain Code=-999 \"cancelled\" UserInfo={NSErrorFailingURLStringKey=https://jsonplaceholder.typicode.com/users/1, NSErrorFailingURLKey=https://jsonplaceholder.typicode.com/users/1, _NSURLErrorRelatedURLSessionTaskErrorKey=(\n    \"LocalDataTask <AA005DD8-5591-4437-86E7-210A92624C73>.<1>\"\n), _NSURLErrorFailingURLSessionTaskErrorKey=LocalDataTask <AA005DD8-5591-4437-86E7-210A92624C73>.<1>, NSLocalizedDescription=cancelled})), request: Hyperspace.HTTP.Request(url: Optional(https://jsonplaceholder.typicode.com/users/1), method: Optional(\"GET\"), headers: Optional([:]), body: nil), response: nil))"

> ```

Reviewer: It could be good to have another section in here talking about the potential tradeoffs/considerations when implementing SSL pinning. For example, what happens when there is a failure or something happens and the certificates change without proper planning. The main thing to highlight would probably be that pinning failures mean that no network requests go through, which would likely provide a "broken" end user experience.

Author: Agreed, I added some additional considerations - let me know if that is similar to what you were thinking.

```

## Pinning Considerations

In addition to the implementation, there are a number of other things to consider when it comes to SSL certificate pinning. When set up correctly, pinning adds an additional layer of security to your network communications. But when set up poorly, it can become an ongoing maintenance nightmare. Here are a few questions to ask as you consider how to implement pinning:

1) Do you _need_ pinning? iOS and TLS are already doing a lot of work for your users behind the scenes. Is the information you are transmitting worth the time and coordination of implementing pinning?

> 1) What happens to a given version of the app when the certificate expires? The server will change its certificate, which will cause pinning to fail. How will you handle this? By default, a pinning failure will cause the request to fail, meaning the user experience of your app will likely be broken. One quick app-side fix for this would be to utilize `DomainConfiguration.expirationPolicy.allow(after:)` functionality, which will automatically stop enforcing pinning to a given domain after a certain date (for example, the certificate's expiration date). Keep in mind that while continually shipping updates to your app containing new certificates is a viable strategy, it effectively gives each version of your app a shelf-life that can not be increased.

Reviewer 4: `1)` shows twice
